File: 백준/Gold/1520. 내리막 길/내리막 길.swift
import Foundation

let input = readLine()!.split(separator:" ").map { Int($0)! }
let m = input[0], n = input[1]   //m: x / n : y
var board:[[Int]] = []
for _ in 0..<m {
  board.append(readLine()!.split(separator:" ").map { Int($0)! })
}

var dp = Array(repeating: Array(repeating: -1, count: n), count:m)

let dx = [-1, 1, 0, 0]
let dy = [0, 0, -1, 1]

var ny = 0, nx = 0

func dfs(_ x: Int, _ y: Int) -> Int {
    if x == m-1 && y == n-1 {
        return 1
    }
    
    if dp[x][y] != -1 {
        return dp[x][y]
    }

    dp[x][y] = 0
    for i in 0..<4 {
      nx = x + dx[i]
      ny = y + dy[i]

      guard 0..<m ~= nx && 0..<n ~= ny else { continue }

      if board[x][y] > board[nx][ny] {
        dp[x][y] += dfs(nx, ny)
      }
    }
    
    return dp[x][y]
}

print(dfs(0, 0))

File: 백준/Gold/1520. 내리막 길/내리막 길.py
import sys
input = sys.stdin.readline

m, n = map(int, input().rstrip().split())
board = [list(map(int, input().rstrip().split())) for _ in range(m)]
dp = [[-1] * n for _ in range(m)]  # 거리를 저장할 DP 테이블

dx = [-1, 1, 0, 0]   # 좌, 우, 하
dy = [0, 0, -1, 1]

def dfs(x, y):
  if x == m-1 and y == n-1:
    return 1

  if dp[x][y] != -1:
    return dp[x][y]

  dp[x][y] = 0
  
  for i in range(4):
    nx = x + dx[i]
    ny = y + dy[i]

    if 0<=nx<m and 0<=ny<n and board[x][y] > board[nx][ny]:
      dp[x][y] += dfs(nx, ny)

  return dp[x][y]

print(dfs(0, 0))